File: quick_sort.cpp
#include <iostream>
using namespace std; 
int partition(int *array, int left, int right){
    int mid = (left + right)/2;
    int pivot = array[mid];
    swap(array[mid], array[left]);
    int i = left + 1;
    int j = right;
    while(i<=j){
        while(i<=j && array[i] <= pivot)
            i++;
        while(i<=j && array[j] > pivot)
            j--;
        if (i<j)
            swap(array[i], array[j]);
    }
    swap(array[left], array[i-1]);
    return i - 1;
}

void quick_sort(int *array, int left, int right){
    if (left<right){
        int part = partition(array, left, right);
        quick_sort(array, left, part-1);
        quick_sort(array, part+1, right);
    }
}
int main()
{
    int array[] = {9,1,8,2,7,3,7,4,6,5};
    int size = 9;
    quick_sort(array, 0, size);
    for (int i = 0; i <=size ; ++i) {
        cout << array[i] << endl;
    }
    return 0;
}

File: quick_sort.py
arr = [1, 1, 8, 5, 6, 7, 3, 4, 2]


def sort(left, right, arr):
    if left >= right:
        return
    pivot = arr[(left + right) // 2]
    index = partition(left, right, arr, pivot)
    sort(left, index - 1, arr)
    sort(index, right, arr)


def partition(left, right, arr, pivot):
    while(left <= right):
        while(arr[left] < pivot):
            left += 1
        while(arr[right] > pivot):
            right -= 1
        if (left <= right):
            arr[left], arr[right] = arr[right], arr[left]
            left, right = left + 1, right - 1
    return left

sort(0, len(arr) - 1, arr)
print arr
